File: backend/app/volume_tags.py
from __future__ import annotations

import math
import re
import subprocess
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .audio_conversion_jobs import creation_flags, resolve_ffmpeg_path
from .database import connect, get_setting, invalidate_library_query_cache, rows_to_dicts
from .file_tags import write_replaygain_tags
from .scanner import path_key


TRACK_VOLUME_COLUMNS = """
    id, path, title, artist, album, album_artist, duration_seconds,
    replaygain_track_gain_db, replaygain_album_gain_db,
    replaygain_track_peak, replaygain_album_peak, genre
"""

REPLAYGAIN_REFERENCE_INTEGRATED = -18.0
I_LINE_RE = re.compile(r"\bI:\s*(-?\d+(?:\.\d+)?)\s*LUFS", re.IGNORECASE)
PEAK_LINE_RE = re.compile(r"\bPeak:\s*(-?\d+(?:\.\d+)?)\s*dB(?:FS|TP)?", re.IGNORECASE)
TRUE_SETTING_VALUES = {"1", "true", "yes", "on"}


@dataclass(frozen=True)
class LoudnessScan:
    integrated: float
    peak: float | None


def parse_ffmpeg_ebur128(output: str) -> LoudnessScan:
    """Read the final FFmpeg ebur128 summary values from stderr."""
    loudness_matches = I_LINE_RE.findall(output)
    if not loudness_matches:
        raise ValueError("FFmpeg did not report integrated loudness")
    peak_matches = PEAK_LINE_RE.findall(output)
    peak = None
    if peak_matches:
        try:
            peak = 10 ** (float(peak_matches[-1]) / 20)
        except ValueError:
            peak = None
    return LoudnessScan(integrated=float(loudness_matches[-1]), peak=peak)


def track_gain_from_loudness(integrated: float) -> float:
    return round(REPLAYGAIN_REFERENCE_INTEGRATED - integrated, 2)


def album_loudness(scans: list[tuple[LoudnessScan, float | None]]) -> float | None:
    if not scans:
        return None
    weighted_energy = 0.0
    total_duration = 0.0
    for scan, duration in scans:
        weight = max(1.0, float(duration or 180.0))
        weighted_energy += weight * (10 ** (scan.integrated / 10))
        total_duration += weight
    if total_duration <= 0 or weighted_energy <= 0:
        return None
    return 10 * math.log10(weighted_energy / total_duration)


def scan_loudness(ffmpeg_path: Path, source: Path, duration_seconds: float | None = None) -> LoudnessScan:
    timeout = max(45, min(300, int(float(duration_seconds or 180.0) * 3 + 30)))
    try:
        result = subprocess.run(
            [
                str(ffmpeg_path),
                "-hide_banner",
                "-nostdin",
                "-nostats",
                "-i",
                str(source),
                "-vn",
                "-sn",
                "-dn",
                "-filter:a",
                "ebur128=peak=true",
                "-f",
                "null",
                "-",
            ],
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=timeout,
            creationflags=creation_flags(),
        )
    except subprocess.TimeoutExpired as exc:
        raise ValueError("FFmpeg loudness scan timed out") from exc
    except OSError as exc:
        raise ValueError(f"Could not start FFmpeg: {exc}") from exc

    output = "\n".join(part for part in [result.stderr, result.stdout] if part)
    if result.returncode != 0 and "Integrated loudness" not in output:
        detail = output.strip().splitlines()[-1] if output.strip() else f"exit code {result.returncode}"
        raise ValueError(f"FFmpeg could not scan this file: {detail}")
    return parse_ffmpeg_ebur128(output)


def selected_volume_tracks(conn, track_ids: list[int] | None, limit: int) -> list[dict[str, Any]]:
    if track_ids:
        unique_ids = list(dict.fromkeys(int(track_id) for track_id in track_ids))
        if not unique_ids:
            return []
        placeholders = ",".join("?" for _ in unique_ids)
        return rows_to_dicts(
            conn.execute(
                f"""
                SELECT {TRACK_VOLUME_COLUMNS}
                FROM tracks
                WHERE id IN ({placeholders})
                ORDER BY lower(coalesce(artist, '')), lower(coalesce(album, '')),
                         coalesce(disc_number, 0), coalesce(track_number, 0),
                         lower(coalesce(title, ''))
                LIMIT ?
                """,
                [*unique_ids, limit],
            )
        )

    return rows_to_dicts(
        conn.execute(
            f"""
            SELECT {TRACK_VOLUME_COLUMNS}
            FROM tracks
            WHERE lower(coalesce(genre, '')) NOT IN ('audiobook', 'podcast')
            ORDER BY datetime(date_added) DESC, id DESC
            LIMIT ?
            """,
            (limit,),
        )
    )


def album_key(track: dict[str, Any]) -> tuple[str, str]:
    artist = str(track.get("album_artist") or track.get("artist") or "").strip().casefold()
    album = str(track.get("album") or "").strip().casefold()
    return artist, album


def value_changed(current: object, proposed: float | None, tolerance: float) -> bool:
    if proposed is None:
        return False
    if current is None:
        return True
    try:
        return abs(float(current) - proposed) > tolerance
    except (TypeError, ValueError):
        return True


def volume_preview(row: dict[str, Any]) -> dict[str, Any]:
    return {
        "track_id": int(row["id"]),
        "path": str(row["path"]),
        "title": row.get("title"),
        "artist": row.get("artist"),
        "album": row.get("album"),
        "current_track_gain_db": row.get("replaygain_track_gain_db"),
        "proposed_track_gain_db": None,
        "current_track_peak": row.get("replaygain_track_peak"),
        "proposed_track_peak": None,
        "current_album_gain_db": row.get("replaygain_album_gain_db"),
        "proposed_album_gain_db": None,
        "current_album_peak": row.get("replaygain_album_peak"),
        "proposed_album_peak": None,
        "changed": False,
        "applied": False,
        "error": None,
    }


def mark_volume_changed(preview: dict[str, Any]) -> bool:
    changed = (
        value_changed(preview["current_track_gain_db"], preview["proposed_track_gain_db"], 0.05)
        or value_changed(preview["current_track_peak"], preview["proposed_track_peak"], 0.0005)
        or value_changed(preview["current_album_gain_db"], preview["proposed_album_gain_db"], 0.05)
        or value_changed(preview["current_album_peak"], preview["proposed_album_peak"], 0.0005)
    )
    preview["changed"] = changed
    return changed


def manual_volume_previews(
    rows: list[dict[str, Any]],
    track_gain_db: float | None,
    track_peak: float | None,
    album_gain_db: float | None,
    album_peak: float | None,
) -> list[dict[str, Any]]:
    previews: list[dict[str, Any]] = []
    for row in rows:
        preview = volume_preview(row)
        preview["proposed_track_gain_db"] = None if track_gain_db is None else round(float(track_gain_db), 2)
        preview["proposed_track_peak"] = None if track_peak is None else round(float(track_peak), 6)
        preview["proposed_album_gain_db"] = None if album_gain_db is None else round(float(album_gain_db), 2)
        preview["proposed_album_peak"] = None if album_peak is None else round(float(album_peak), 6)
        mark_volume_changed(preview)
        previews.append(preview)
    return previews


def applied_value(preview: dict[str, Any], proposed_key: str, current_key: str, manual_mode: bool) -> float | None:
    proposed = preview.get(proposed_key)
    if proposed is not None or not manual_mode:
        return proposed if isinstance(proposed, (int, float)) else None
    current = preview.get(current_key)
    return current if isinstance(current, (int, float)) else None


def build_volume_tag_response(
    track_ids: list[int] | None,
    mode: str,
    apply: bool,
    write_to_file: bool | None,
    limit: int,
    manual_track_gain_db: float | None = None,
    manual_track_peak: float | None = None,
    manual_album_gain_db: float | None = None,
    manual_album_peak: float | None = None,
) -> dict[str, Any]:
    with connect() as conn:
        should_write_to_file = bool(write_to_file) if write_to_file is not None else get_setting(conn, "write_ratings_to_files") in TRUE_SETTING_VALUES
        rows = selected_volume_tracks(conn, track_ids, limit)
        errors: list[str] = []
        ffmpeg_path: Path | None = None
        checked_paths: list[str] = []

        if mode == "manual":
            if all(value is None for value in [manual_track_gain_db, manual_track_peak, manual_album_gain_db, manual_album_peak]):
                errors.append("Enter at least one manual gain or peak value.")
            previews = manual_volume_previews(rows, manual_track_gain_db, manual_track_peak, manual_album_gain_db, manual_album_peak)
        else:
            ffmpeg_path, _configured, candidates = resolve_ffmpeg_path(conn)
            checked_paths = [str(candidate) for candidate in candidates[:12]]
            if ffmpeg_path is None:
                return {
                    "total": 0,
                    "changed": 0,
                    "applied": 0,
                    "errors": [
                        "FFmpeg is required for volume tag analysis. Install it from File Management > Audio Conversion."
                    ],
                    "previews": [],
                    "ffmpeg_path": None,
                    "checked_paths": checked_paths,
                }
            previews = []
            group_scans: dict[tuple[str, str], list[tuple[LoudnessScan, float | None]]] = {}

            for row in rows:
                source = Path(str(row["path"]))
                preview = volume_preview(row)
                if not source.exists():
                    preview["error"] = "File is missing on disk"
                    errors.append(f"{source}: file is missing on disk")
                    previews.append(preview)
                    continue
                try:
                    scan = scan_loudness(ffmpeg_path, source, row.get("duration_seconds"))
                except ValueError as exc:
                    message = str(exc)
                    preview["error"] = message
                    errors.append(f"{source.name}: {message}")
                    previews.append(preview)
                    continue
                group_scans.setdefault(album_key(row), []).append((scan, row.get("duration_seconds")))
                preview["proposed_track_gain_db"] = track_gain_from_loudness(scan.integrated)
                preview["proposed_track_peak"] = scan.peak
                previews.append(preview)

            album_values: dict[tuple[str, str], tuple[float | None, float | None]] = {}
            for key, scans in group_scans.items():
                loudness = album_loudness(scans)
                album_gain = round(REPLAYGAIN_REFERENCE_INTEGRATED - loudness, 2) if loudness is not None else None
                peaks = [scan.peak for scan, _duration in scans if scan.peak is not None]
                album_values[key] = (album_gain, max(peaks) if peaks else None)

            row_by_id = {int(row["id"]): row for row in rows}
            for preview in previews:
                row = row_by_id.get(int(preview["track_id"]))
                if row is None or preview.get("error"):
                    continue
                proposed_album_gain, proposed_album_peak = album_values.get(album_key(row), (None, None))
                preview["proposed_album_gain_db"] = proposed_album_gain
                preview["proposed_album_peak"] = proposed_album_peak
                mark_volume_changed(preview)

        if errors and not previews:
            return {
                "total": 0,
                "changed": 0,
                "applied": 0,
                "errors": errors[:100],
                "previews": [],
                "ffmpeg_path": str(ffmpeg_path) if ffmpeg_path else None,
                "checked_paths": checked_paths,
            }

        applied = 0
        row_by_id = {int(row["id"]): row for row in rows}
        for preview in previews:
            track_id = int(preview["track_id"])
            row = row_by_id.get(track_id)
            if row is None or preview.get("error"):
                continue
            if not apply or not preview.get("changed"):
                continue

            source = Path(str(row["path"]))
            applied_track_gain = applied_value(preview, "proposed_track_gain_db", "current_track_gain_db", mode == "manual")
            applied_track_peak = applied_value(preview, "proposed_track_peak", "current_track_peak", mode == "manual")
            applied_album_gain = applied_value(preview, "proposed_album_gain_db", "current_album_gain_db", mode == "manual")
            applied_album_peak = applied_value(preview, "proposed_album_peak", "current_album_peak", mode == "manual")
            file_modified_at = None
            if should_write_to_file:
                try:
                    write_replaygain_tags(
                        source,
                        applied_track_gain,
                        applied_track_peak,
                        applied_album_gain,
                        applied_album_peak,
                    )
                    if source.exists():
                        file_modified_at = datetime.fromtimestamp(source.stat().st_mtime, timezone.utc).replace(microsecond=0).isoformat()
                except (OSError, ValueError) as exc:
                    message = str(exc)
                    preview["error"] = message
                    errors.append(f"{source.name}: {message}")
                    continue

            conn.execute(
                """
                UPDATE tracks
                SET replaygain_track_gain_db = ?,
                    replaygain_track_peak = ?,
                    replaygain_album_gain_db = ?,
                    replaygain_album_peak = ?,
                    file_modified_at = coalesce(?, file_modified_at),
                    updated_at = datetime('now')
                WHERE id = ?
                """,
                (
                    applied_track_gain,
                    applied_track_peak,
                    applied_album_gain,
                    applied_album_peak,
                    file_modified_at,
                    track_id,
                ),
            )
            conn.execute("DELETE FROM track_metadata_cache WHERE path_key = ?", (path_key(source),))
            preview["applied"] = True
            applied += 1

        if applied:
            invalidate_library_query_cache(conn)
        conn.commit()

    return {
        "total": len(previews),
        "changed": sum(1 for preview in previews if preview.get("changed")),
        "applied": applied,
        "errors": errors[:100],
        "previews": previews,
        "ffmpeg_path": str(ffmpeg_path) if ffmpeg_path else None,
        "checked_paths": checked_paths if ffmpeg_path is None else [],
    }

File: src-tauri/src/library/volume_tags.rs
use super::*;

const REPLAYGAIN_REFERENCE_INTEGRATED: f64 = -18.0;

#[derive(Clone, Copy)]
struct LoudnessScan {
    integrated: f64,
    peak: Option<f64>,
}

fn volume_value_changed(current: Option<f64>, proposed: Option<f64>, tolerance: f64) -> bool {
    match proposed {
        None => false,
        Some(proposed) if !proposed.is_finite() => false,
        Some(proposed) => current
            .map(|current| (current - proposed).abs() > tolerance)
            .unwrap_or(true),
    }
}

fn volume_changed(preview: &DesktopVolumeTagPreview) -> bool {
    volume_value_changed(
        preview.current_track_gain_db,
        preview.proposed_track_gain_db,
        0.05,
    ) || volume_value_changed(
        preview.current_track_peak,
        preview.proposed_track_peak,
        0.0005,
    ) || volume_value_changed(
        preview.current_album_gain_db,
        preview.proposed_album_gain_db,
        0.05,
    ) || volume_value_changed(
        preview.current_album_peak,
        preview.proposed_album_peak,
        0.0005,
    )
}

fn parse_ffmpeg_ebur128(output: &str) -> Result<LoudnessScan, String> {
    let loudness_re = regex::Regex::new(r"(?i)\bI:\s*(-?\d+(?:\.\d+)?)\s*LUFS")
        .map_err(|error| format!("Could not build loudness parser: {error}"))?;
    let peak_re = regex::Regex::new(r"(?i)\bPeak:\s*(-?\d+(?:\.\d+)?)\s*dB(?:FS|TP)?")
        .map_err(|error| format!("Could not build peak parser: {error}"))?;
    let integrated = loudness_re
        .captures_iter(output)
        .filter_map(|captures| captures.get(1).and_then(|m| m.as_str().parse::<f64>().ok()))
        .last()
        .ok_or_else(|| "FFmpeg did not report integrated loudness".to_string())?;
    let peak = peak_re
        .captures_iter(output)
        .filter_map(|captures| captures.get(1).and_then(|m| m.as_str().parse::<f64>().ok()))
        .last()
        .map(|db| 10f64.powf(db / 20.0));
    Ok(LoudnessScan { integrated, peak })
}

fn scan_loudness(ffmpeg_path: &Path, track: &DesktopTrack) -> Result<LoudnessScan, String> {
    let source = Path::new(&track.path);
    if !source.exists() || !source.is_file() {
        return Err("File is missing on disk".to_string());
    }
    let mut command = Command::new(ffmpeg_path);
    command.args([
        "-hide_banner",
        "-nostdin",
        "-nostats",
        "-i",
        &track.path,
        "-vn",
        "-sn",
        "-dn",
        "-filter:a",
        "ebur128=peak=true",
        "-f",
        "null",
        "-",
    ]);
    #[cfg(windows)]
    command.creation_flags(CREATE_NO_WINDOW);
    let output = command
        .output()
        .map_err(|error| format!("Could not start FFmpeg: {error}"))?;
    let text = format!(
        "{}\n{}",
        String::from_utf8_lossy(&output.stderr),
        String::from_utf8_lossy(&output.stdout)
    );
    if !output.status.success() && !text.contains("Integrated loudness") {
        let detail = text
            .lines()
            .rev()
            .map(str::trim)
            .find(|line| !line.is_empty())
            .unwrap_or("FFmpeg failed");
        return Err(format!("FFmpeg could not scan this file: {detail}"));
    }
    parse_ffmpeg_ebur128(&text)
}

fn track_gain_from_loudness(integrated: f64) -> f64 {
    round_to(REPLAYGAIN_REFERENCE_INTEGRATED - integrated, 2)
}

fn album_loudness(scans: &[(LoudnessScan, Option<f64>)]) -> Option<f64> {
    if scans.is_empty() {
        return None;
    }
    let mut weighted_energy = 0.0;
    let mut total_duration = 0.0;
    for (scan, duration) in scans {
        let weight = duration.unwrap_or(180.0).max(1.0);
        weighted_energy += weight * 10f64.powf(scan.integrated / 10.0);
        total_duration += weight;
    }
    (total_duration > 0.0 && weighted_energy > 0.0)
        .then(|| 10.0 * (weighted_energy / total_duration).log10())
}

fn volume_album_key(track: &DesktopTrack) -> (String, String) {
    (
        track
            .album_artist
            .as_deref()
            .or(track.artist.as_deref())
            .unwrap_or("")
            .trim()
            .to_lowercase(),
        track.album.as_deref().unwrap_or("").trim().to_lowercase(),
    )
}

fn round_to(value: f64, places: i32) -> f64 {
    let factor = 10f64.powi(places);
    (value * factor).round() / factor
}

fn applied_volume_value(
    proposed: Option<f64>,
    current: Option<f64>,
    manual_mode: bool,
) -> Option<f64> {
    if proposed.is_some() || !manual_mode {
        proposed
    } else {
        current
    }
}

#[tauri::command]
pub fn volume_tags_preview(
    _state: State<'_, DesktopLibraryState>,
    track_ids: Option<Vec<i64>>,
    mode: Option<String>,
    write_to_file: Option<bool>,
    manual_track_gain_db: Option<f64>,
    manual_track_peak: Option<f64>,
    manual_album_gain_db: Option<f64>,
    manual_album_peak: Option<f64>,
    apply: Option<bool>,
    limit: Option<usize>,
) -> Result<DesktopVolumeTagResponse, String> {
    let connection = open_database()?;
    let mode = mode.unwrap_or_else(|| "manual".to_string());
    let manual_mode = mode.eq_ignore_ascii_case("manual");
    let tracks = select_tracks_by_ids_or_limit(
        &connection,
        track_ids,
        limit
            .unwrap_or(if manual_mode { 200 } else { 50 })
            .clamp(1, 10_000),
    )?;
    let mut previews = Vec::new();
    let apply = apply.unwrap_or(false);
    let should_write_to_file = write_to_file
        .unwrap_or_else(|| truthy_setting(&connection, "write_ratings_to_files", false));
    let mut applied = 0i64;
    let mut errors = Vec::new();
    let mut ffmpeg_path = None;
    let mut checked_paths = Vec::new();

    if manual_mode
        && manual_track_gain_db.is_none()
        && manual_track_peak.is_none()
        && manual_album_gain_db.is_none()
        && manual_album_peak.is_none()
    {
        errors.push("Enter at least one manual gain or peak value.".to_string());
    }

    for track in &tracks {
        previews.push(DesktopVolumeTagPreview {
            track_id: track.id,
            path: track.path.clone(),
            title: track.title.clone(),
            artist: track.artist.clone(),
            album: track.album.clone(),
            current_track_gain_db: track.replaygain_track_gain_db,
            proposed_track_gain_db: None,
            current_track_peak: track.replaygain_track_peak,
            proposed_track_peak: None,
            current_album_gain_db: track.replaygain_album_gain_db,
            proposed_album_gain_db: None,
            current_album_peak: track.replaygain_album_peak,
            proposed_album_peak: None,
            changed: false,
            applied: false,
            error: None,
        });
    }

    if manual_mode {
        for preview in &mut previews {
            preview.proposed_track_gain_db = manual_track_gain_db.map(|value| round_to(value, 2));
            preview.proposed_track_peak = manual_track_peak.map(|value| round_to(value, 6));
            preview.proposed_album_gain_db = manual_album_gain_db.map(|value| round_to(value, 2));
            preview.proposed_album_peak = manual_album_peak.map(|value| round_to(value, 6));
            preview.changed = volume_changed(preview);
        }
    } else {
        let (resolved_ffmpeg, candidates) = tools::resolved_ffmpeg_tool_path()?;
        checked_paths = candidates
            .iter()
            .take(12)
            .map(|path| path.to_string_lossy().to_string())
            .collect();
        let Some(path) = resolved_ffmpeg else {
            return Ok(DesktopVolumeTagResponse {
                total: 0,
                changed: 0,
                applied: 0,
                errors: vec![
                    "FFmpeg is required for volume tag analysis. Install it from File Management > Optional Dependencies."
                        .to_string(),
                ],
                previews: Vec::new(),
                ffmpeg_path: None,
                checked_paths,
            });
        };
        ffmpeg_path = Some(path.to_string_lossy().to_string());
        let mut group_scans: HashMap<(String, String), Vec<(LoudnessScan, Option<f64>)>> =
            HashMap::new();
        let mut scans_by_id: HashMap<i64, LoudnessScan> = HashMap::new();
        for (index, track) in tracks.iter().enumerate() {
            match scan_loudness(&path, track) {
                Ok(scan) => {
                    scans_by_id.insert(track.id, scan);
                    group_scans
                        .entry(volume_album_key(track))
                        .or_default()
                        .push((scan, track.duration_seconds));
                    previews[index].proposed_track_gain_db =
                        Some(track_gain_from_loudness(scan.integrated));
                    previews[index].proposed_track_peak = scan.peak.map(|value| round_to(value, 6));
                }
                Err(error) => {
                    previews[index].error = Some(error.clone());
                    errors.push(format!(
                        "{}: {error}",
                        Path::new(&track.path)
                            .file_name()
                            .and_then(|value| value.to_str())
                            .unwrap_or(&track.path)
                    ));
                }
            }
        }
        let album_values = group_scans
            .iter()
            .map(|(key, scans)| {
                let gain = album_loudness(scans)
                    .map(|loudness| round_to(REPLAYGAIN_REFERENCE_INTEGRATED - loudness, 2));
                let peak = scans
                    .iter()
                    .filter_map(|(scan, _)| scan.peak)
                    .reduce(f64::max)
                    .map(|value| round_to(value, 6));
                (key.clone(), (gain, peak))
            })
            .collect::<HashMap<_, _>>();
        for (index, track) in tracks.iter().enumerate() {
            if !scans_by_id.contains_key(&track.id) {
                continue;
            }
            let (album_gain, album_peak) = album_values
                .get(&volume_album_key(track))
                .copied()
                .unwrap_or((None, None));
            previews[index].proposed_album_gain_db = album_gain;
            previews[index].proposed_album_peak = album_peak;
            previews[index].changed = volume_changed(&previews[index]);
        }
    }

    let track_by_id = tracks
        .iter()
        .map(|track| (track.id, track))
        .collect::<HashMap<_, _>>();
    for preview in &mut previews {
        if apply && preview.changed {
            let Some(track) = track_by_id.get(&preview.track_id) else {
                continue;
            };
            let applied_track_gain = applied_volume_value(
                preview.proposed_track_gain_db,
                preview.current_track_gain_db,
                manual_mode,
            );
            let applied_track_peak = applied_volume_value(
                preview.proposed_track_peak,
                preview.current_track_peak,
                manual_mode,
            );
            let applied_album_gain = applied_volume_value(
                preview.proposed_album_gain_db,
                preview.current_album_gain_db,
                manual_mode,
            );
            let applied_album_peak = applied_volume_value(
                preview.proposed_album_peak,
                preview.current_album_peak,
                manual_mode,
            );
            let mut file_modified_at = None;
            if should_write_to_file {
                let path = Path::new(&track.path);
                match metadata::write_replaygain_tags(
                    path,
                    applied_track_gain,
                    applied_track_peak,
                    applied_album_gain,
                    applied_album_peak,
                ) {
                    Ok(()) => {
                        file_modified_at = metadata::modified_time_iso(path);
                    }
                    Err(error) => {
                        let message = format!("{}: {error}", preview.path);
                        preview.error = Some(message.clone());
                        errors.push(message);
                        continue;
                    }
                }
            }
            match connection.execute(
                r#"
                UPDATE tracks
                SET replaygain_track_gain_db = ?,
                    replaygain_track_peak = ?,
                    replaygain_album_gain_db = ?,
                    replaygain_album_peak = ?,
                    file_modified_at = coalesce(?, file_modified_at),
                    updated_at = datetime('now')
                WHERE id = ?
                "#,
                params![
                    applied_track_gain,
                    applied_track_peak,
                    applied_album_gain,
                    applied_album_peak,
                    file_modified_at,
                    preview.track_id
                ],
            ) {
                Ok(_) => {
                    preview.applied = true;
                    applied += 1;
                    let _ = connection.execute(
                        "DELETE FROM track_metadata_cache WHERE path_key = ?",
                        params![normalized_path_key(&track.path)],
                    );
                }
                Err(error) => {
                    let message = format!("{}: {error}", preview.path);
                    preview.error = Some(message.clone());
                    errors.push(message);
                }
            }
        }
    }
    if applied > 0 {
        clear_library_query_cache(&connection);
    }
    let changed = previews.iter().filter(|preview| preview.changed).count() as i64;
    let has_ffmpeg_path = ffmpeg_path.is_some();
    Ok(DesktopVolumeTagResponse {
        total: previews.len() as i64,
        changed,
        applied,
        errors,
        previews,
        ffmpeg_path,
        checked_paths: if has_ffmpeg_path {
            Vec::new()
        } else {
            checked_paths
        },
    })
}
